File: Exercises/ClassicalMechanics/Orbital/Python/orbital.py
############################################################
# Martian Invasion
############################################################

# Import matplotlib library for plotting
import matplotlib
matplotlib.use('Agg') # this is a hack to keep matplotlib
                      # non-interactive/keep python from complaining
                      # if you're using an ssh connection
import matplotlib.pyplot as pyplot

# Import numpy for numerical calculations, vectors, etc.
import numpy as np
from numpy.linalg import norm as norm

# Import our parameter file, the '*' says take everything in that
# file, and make it available for use in this file
from params import *

# Calculates the force of gravity given displacement vectors r1 & r1
# and scalar masses m1, m2
def force_gravity(r1, r2, m1, m2):
    
    # numpy can subtract arrays like vectors
    r = r1 - r2
    F = - G*m1*m2*norm(r)**(-3.)*r
    return F

# Sum the forces on each body in the current system
def calculate_forces(positions, masses):
    
    # set n to be the number of bodies to consider
    n = len(positions)
    
    # create an array of 2D vectors of length n
    forces = np.zeros(shape=(n,2))

    # for each body, add the gravitational force to each other body
    for i in range(n):
        for j in range(n):
            # make sure we are not calculating the gravity to itself
            if i != j:
                forces[i] += force_gravity(positions[i], positions[j], 
                                           masses[i], masses[j])
                
    # forces[3] += np.array([ -1e13, 1e13 ])

    return forces
    
# Update the positions using velocity Verlet integration
def update_pos(positions, velocities, masses):

    # calculate the total force and accelerations on each body
    forces = calculate_forces(positions, masses)
    accel = np.array([forces[i]/masses[i] for i in range(len(masses))])

    # update the positions using the verlet method
    positions += velocities*dt + .5*accel*dt**2
    
    # recalculate the accelerations for the new positions
    forces = calculate_forces(positions, masses)
    newaccel = np.array([forces[i]/masses[i] for i in range(len(masses))])

    # use average acceleration to update velocities
    velocities += .5*(newaccel + accel)*dt

    return positions, velocities

def plot_radii(trajectories):
    # settings for plotting
    IMAGE_PATH = "radii.png"
    TITLE = "Rocket Projectile"
    YAXIS = "Y"
    XAXIS = "X"

    # Plot the trajectories
    print "Plotting radii."
    plt = pyplot.figure(figsize=(15, 10), dpi=80, facecolor='w')
    ax = pyplot.axes()
    ax.set_xlabel(XAXIS) 
    ax.set_ylabel(YAXIS)
    ax.set_title(TITLE)

    R_M = np.sqrt(trajectories[:,2][:, 0]**2 + trajectories[:,2][:, 1]**2)
    R_R = np.sqrt(trajectories[:,3][:, 0]**2 + trajectories[:,3][:, 1]**2)
    R_E = np.sqrt(trajectories[:,1][:, 0]**2 + trajectories[:,1][:, 1]**2)
                  
    X = range(len(R_M))

    ax.plot(X, R_M, "-", alpha=.7, linewidth=3, label="Mars")
    ax.plot(X, R_R, "-", alpha=.7, linewidth=3, label="Rocket")
    ax.plot(X, R_E, "-", alpha=.7, linewidth=3, label="Earth")
    
    ax.legend(bbox_to_anchor=(1., 1.), loc="best", 
              ncol=1, fancybox=True, shadow=True)

    # Save our plot
    print "Saving plot to %s." % IMAGE_PATH
    plt.savefig(IMAGE_PATH, bbox_inches='tight')
    

def plot_orbit(trajectories):

    # settings for plotting
    IMAGE_PATH = "trajectories.png"
    TITLE = "Rocket Projectile"
    YAXIS = "Y"
    XAXIS = "X"

    # Plot the trajectories
    print "Plotting."
    plt = pyplot.figure(figsize=(15, 10), dpi=80, facecolor='w')
    ax = pyplot.axes()
    ax.set_xlabel(XAXIS) 
    ax.set_ylabel(YAXIS)
    ax.set_title(TITLE)

    # Trajectories: trace trajectory of objects
    ax.plot(trajectories[:,0][:, 0], trajectories[:,0][:, 1], 
            "yo-", alpha=.7, linewidth=3, label="Sun")
    ax.plot(trajectories[:,1][:, 0], trajectories[:,1][:, 1], 
            "b-", alpha=.7, linewidth=3, label="Earth")
    ax.plot(trajectories[:,2][:, 0], trajectories[:,2][:, 1], 
            "r-", alpha=.7, linewidth=3, label="Mars")
    ax.plot(trajectories[:,3][:, 0], trajectories[:,3][:, 1], 
            "g-", alpha=.7, linewidth=3, label="Rocket")

    # Objects: draw a dot on the last trajectory point
    ax.plot(trajectories[-1,1,0], trajectories[-1,1,1], "bo-")
    ax.plot(trajectories[-1,2,0], trajectories[-1,2,1], "ro-")
    ax.plot(trajectories[-1,3,0], trajectories[-1,3,1], "go-")

    pyplot.xlim(-3e11, 3e11)
    pyplot.ylim(-3e11, 3e11)
    # pyplot.axis('equal')

    ax.legend(bbox_to_anchor=(1., 1.), loc="best", 
              ncol=1, fancybox=True, shadow=True)

    # Save our plot
    print "Saving plot to %s." % IMAGE_PATH
    plt.savefig(IMAGE_PATH, bbox_inches='tight')

def calculate_trajectory(V_Y, THETA, ADJUSTMENT):
    
    V_Y += ADJUSTMENT

    # positions is an array of vectors in a 2D plane specifying the
    # positions of [sun, earth, mars, rocket] in m from origin
    positions = np.array([
            [0, 0], 
            [RADIUS_E*np.cos(THETA), -RADIUS_E*np.sin(THETA)],
            [RADIUS_M, 0],
            [R1, 0],
            ])

    # velocities of each body in m/s
    velocities = np.array([
            [0, 0],
            [VEL_E*np.sin(THETA), VEL_E*np.cos(THETA)],
            [0, VEL_M],
            [0, VEL_M - VEL_R]
            # [0, VEL_M]
            ])

    print "Applying primary Hohmann boost."
    velocities[3,1] += V_Y
    
    # masses of each body in kg
    masses = np.array([MASS_S, MASS_E, MASS_M, MASS_R])
    
    BOOSTED = False

    trajectories = []
    print "Calculating trajectory."
    for t in range(MAX_STEP):
        # we append 'np.array(positions)' to the list so we make a
        # COPY of the position array if we just appended 'positions'
        # then we would only append a reference to the array, whose
        # values would be updated each iteration!
        
        ROCKET_Y = positions[3,1]
        trajectories.append(np.array(positions))

        positions, velocities = update_pos(positions, velocities, masses)
        
        if (norm(positions[1] - positions[3]) < DIAM_E/2):
            print "LANDED"
            trajectories = np.array(trajectories)
            return trajectories

        if (ROCKET_Y > 0 and positions[3,1] < 0 and not BOOSTED):
            BOOSTED = True
            print "Applying secondary Hohmann boost."
            velocities[3,1] -= V_Y2

    # turn our python list into a numpy array
    trajectories = np.array(trajectories)
    return trajectories

# This is the function that gets called when we run the program    
def main():

    print "Starting calculation."

    THETA      = np.pi*.45
    ADJUSTMENT = 8.10493e2

    trajectories = calculate_trajectory(V_Y, THETA, ADJUSTMENT)

    plot_orbit(trajectories)
    plot_radii(trajectories)

# This is Python syntax which tells Python to call the function we
# created, called 'main()', only if this file was run directly, rather
# than with 'import orbital'
if __name__ == "__main__":
    main()

File: Exercises/ClassicalMechanics/Orbital/C/orbital.c
// Martian Invasion

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define DIM 2
#define N_BODY 4

#define MAX_T    2.43e7
#define dt       3.6e3    // time step of 1 hour
#define MAX_STEP (int)(MAX_T/dt)
#define G        6.67e-11 // Gravitational constant [N m^2/kg^2]
#define PI 3.14159265359

#define DIAM_E   1.273e7  // Diameter of the earth
#define RADIUS_E 1.521e11 // Earth to sun           [meters]
#define RADIUS_M 2.296e11 // mars to sun            [meters]
#define RADIUS_R 9.281e6  // Mars to rocket orbit   [meters]

#define MASS_S   1.989e30 // Mass of the Sun        [kg]
#define MASS_E   5.972e24 // Mass of Earth          [kg]
#define MASS_M   6.417e23 // Mass of Mars           [kg]
#define MASS_R   1.072e6  // Mass of rocket         [kg]

#define VEL_E    2.98e4   // Velocity of Earth      [m/s]
#define VEL_M    2.41e4   // Velocity of Mars       [m/s]
#define VEL_R    2.14e3   // Velocity of rocket     [m/s]

const double R1 = RADIUS_M-RADIUS_R;


double norm(double r[DIM])
{
    double sqnorm = 0;
    for (int i = 0; i < DIM; i++)
	sqnorm += r[i]*r[i];
    return sqrt(sqnorm);
}

// Calculates the force of gravity given displacement vectors r1 & r1
// and scalar masses m1, m2
void force_gravity(double r1[DIM], double r2[DIM], 
		   double m1, double m2, double F[DIM])
{
    double r[DIM];
    for (int i = 0; i < DIM; i++)
	r[i] = r1[i] - r2[i];
    for (int i = 0; i < DIM; i++)
	F[i] += - G*m1*m2*pow(norm(r), -1)*r[i];
    return;
}

// Sum the forces on each body in the current system
void calculate_forces(double positions[N_BODY][DIM], 
		      double masses[N_BODY], 
		      double forces[N_BODY][DIM])
{
    
    // for each body, add the gravitational force to each other body
    for (int i = 0; i < N_BODY; i++){
        for (int j = 0; j < N_BODY; j++){

            // make sure we are not calculating the gravity to itself
            if (i != j){
		// add the force of gravity to the array of forces
		force_gravity(positions[i], positions[j], 
			      masses[i], masses[j], forces[i]);

	    }
	}
    }

    return;
}

    
// Update the positions using velocity Verlet integration
void update_pos(double positions[N_BODY][DIM], 
		double velocities[N_BODY][DIM], 
		double masses[N_BODY])
{

    // calculate the total force and accelerations on each body
    double forces[N_BODY][DIM], new_forces[N_BODY][DIM];
    double accel[N_BODY][DIM], new_accel[N_BODY][DIM];

    for (int i = 0; i < N_BODY; i++){
	for (int j = 0; j < DIM; j++){
	    forces[i][j] = 0;
	    accel[i][j] = 0;
	    new_forces[i][j] = 0;
	    new_accel[i][j] = 0;
	}
    }

    calculate_forces(positions, masses, forces);

    // update the positions using the verlet method
    for (int i = 0; i < N_BODY; i++){
	for (int j = 0; j < DIM; j++){
	    accel[i][j] = forces[i][j]/masses[i];
	    positions[i][j] += velocities[i][j]*dt + .5*accel[i][j]*dt*dt;
	}
    }
    
    // recalculate the accelerations for the new positions
    calculate_forces(positions, masses, new_forces);

    // use average acceleration to update velocities
    for (int i = 0; i < N_BODY; i++){
	for (int j = 0; j < DIM; j++){
	    new_accel[i][j] = forces[i][j]/masses[i];
	    velocities[i][j] += .5*(new_accel[i][j] + accel[i][j])*dt;
	}
    }

    return;
}

void calculate_trajectory(double V_Y, 
			  double V_Y2, 
			  double THETA, 
			  double ADJUSTMENT)
{
    V_Y += ADJUSTMENT;

    // positions is an array of vectors in a 2D plane specifying the
    // positions of [sun, earth, mars, rocket] in m from origin

    double positions[N_BODY][DIM], velocities[N_BODY][DIM];

    /* Sun */
    positions[0][0] = 0.0;
    positions[0][1] = 0.0;
    velocities[0][0] = 0.0;
    velocities[0][1] = 0.0;


    /* Earth */
    positions[1][0] = RADIUS_E*cos(THETA);
    positions[1][1] = -RADIUS_E*sin(THETA);
    velocities[1][0] = VEL_E*sin(THETA);
    velocities[1][1] = VEL_E*cos(THETA);

    /* Mars */
    positions[2][0] = RADIUS_M;
    positions[2][1] = 0;
    velocities[2][0] = 0;
    velocities[2][1] = VEL_M;

    /* Rocket */
    positions[3][0] = R1;
    positions[3][1] = 0;
    velocities[3][0] = 0;
    velocities[3][1] = VEL_M - VEL_R;

    printf("Applying primary Hohmann boost.\n");
    velocities[3][1] += V_Y;
    
    // masses of each body in kg
    double masses[] = {MASS_S, MASS_E, MASS_M, MASS_R};
    
    int BOOSTED = 0;

    printf("Calculating trajectory.\n");

    for (int i = 0; i < MAX_STEP; i++){

	double ROCKET_Y = positions[3][1];

        update_pos(positions, velocities, masses);
        
	double disp_to_earth[DIM];
	for (int i = 0; i < DIM; i++)
	    disp_to_earth[i] = positions[1][i] - positions[3][i];

	printf("%.2e\t%.2e\n", disp_to_earth[0], disp_to_earth[1]);
        if (norm(disp_to_earth) < DIAM_E/2){
            printf("LANDED\n");
	    return;
	}

        if (ROCKET_Y > 0 && positions[3][1] < 0 && !BOOSTED){
            BOOSTED = 1;
            printf("Applying secondary Hohmann boost.\n");
            velocities[3][1] -= V_Y2;
	}
    }

    return;

}


// This is the function that gets called when we run the program    
int main(int argc, char *argv[])
{
    printf("Starting calculation.\n");

    /* Use analytical solution to find Hohmann velocity difference */
    double V_Y = sqrt(G*MASS_S/R1)*(sqrt(2*RADIUS_E/(RADIUS_E+R1))-1);

    // correction for numerical error of integration techinique
    V_Y *= .97;

    // Secondary Hohmann boost
    double V_Y2 = sqrt(G*MASS_S/RADIUS_E)*(1-sqrt(2*R1/(RADIUS_E+R1)));

    // Parameters for gravity well adjustment
    double ADJUSTMENT = 8.10493e2;

    // Relative orbital angle between Mars and Earth
    double THETA      = PI*.45;

    // Calculate the trajectories
    calculate_trajectory(V_Y, V_Y2, THETA, ADJUSTMENT);

}
